README.md
Action 53
=========

The Action 53 ROM builder chews up a bunch of ROMs and spits out a
multicart.

Overview
--------
The menu has up to five pages, each represented by a tab at the top
of the screen, and each page may have up to 20 activities.  From this
list, the player can view a more detailed description including basic
play instructions.

For each ROM added to the collection, the ROM builder patches each
bank with a small piece of code that allows the activity to reset
back to the menu.  NROM games use one of two patches: a 10-byte one
that fits at $BFF0-$BFF9 or $FFF0-$FFF9, or a 20-byte one that fits
anywhere in the PRG bank.  (Larger games use more involved patches
that change more of the mapper's state.)  Then it stores the ROM's
PRG data, its CHR data compressed with PB53, and information used
by the extractor to reverse the reset patch.

The config file may specify ranges of addresses within $8000-$FFFF
that are not used by the ROM.  The reset patch is placed here, and
other programs' CHR ROM and screenshot tile data may be placed here.
For an NROM-128 program, this unused area automatically includes the
half of $8000-$BFFF where the reset vector does not point.

For each activity, the builder stores information needed to
display it in the menu:  title, author, year of publication, number
of players, a description with up to 16 lines of play instructions,
and a screenshot as defined below.  It also stores information
needed to start the activity after the user has chosen it: which PRG
and CHR banks it uses and where in the program to start running.

A ROM may have multiple entry points within one PRG bank, one for
each activity.  This allows a ROM with multiple separate activities
to become several items in the menu.  Examples of these from the
commercial NES library include _Duck Hunt_ ("Duck Hunt" and "Clay
Shooting") and _Donkey Kong Classics_ ("Donkey Kong" and
"Donkey Kong Jr.").  In fact, packing activities into "submultis",
or ROMs with multiple activities sharing a PRG bank, is the most
efficient way to fit a large number of activities into a collection.
If no entry point is specified, the builder uses the reset vector
from the ROM.

A ROM may have multiple CHR banks.  If you specify one, the menu will
load it into CHR RAM before launching your activity.  Some developers
have used defective startup code that tries to clear the nametables
($2000-$27FF), but it loops too much and ends up clearing the pattern
tables too, so don't do that.  The builder provides a patching
mechanism that allows the config file to fix minor bugs, such as
scribbling over the pattern tables, switching CHR banks, or a wrong
version number.

ROM requirements
----------------
The builder produces a ROM, usually 128 to 2048 KiB.  It can be
configured to use an extended-BNROM mapper (iNES #34) with vertical
mirroring or a custom Action 53 mapper (iNES #28).  Each ROM inserted
into a BNROM collection:

* MUST work with vertical mirroring.
* MUST initialize all hardware and memory needed to run from each
  activity's declared entry point.
* MUST NOT write to $8000-$FFFF after the declared entry point.
* MUST NOT blindly overwrite CHR data if using CHR ROM.
* MUST declare an unused range of the ROM to hold the reset patch
  if the PRG ROM is at least 32768 bytes.
* MUST NOT play samples if the PRG ROM is 16384 bytes or smaller
  and has a reset vector in $8000-$BFFF.

Most of these restrictions are lifted for collections using the
Action 53 mapper, which may use common discrete mappers: UNROM
(iNES #2 or #180), CNROM (iNES #3), AOROM (iNES #7), or BNROM.
But if a game actually uses CNROM, as opposed to being a submulti
with a separate CHR ROM bank for each activity, the cartridge will
need to have 32 KiB of CHR RAM instead of 8 KiB.

Preparing graphics
------------------
A collection needs a title screen and a screenshot for each activity.

The title screen is 30 rows of 32 tiles, each 8 by 8 pixels.
A picture can use up to 256 different tiles. Each 2 by 2 tile
(16 by 16 pixel) area can use one of four color sets in the palette.
Each color set has three colors plus a shared backdrop color.

To save ROM space, text may be overlaid on top of the title screen.
Each 8x8-pixel area covered by text reduces available tiles by 1.
For example, three 128-pixel lines of text take 16 tiles each,
leaving 208 for graphics.

The title screen's palette is specified with a 32-nibble string
representing a hex dump of NES palette memory ($3F00-$3F0F).
The following example specifies a black ($0F) backdrop, color set 0
as grays ($00, $10, $20), color set 1 as reds ($06, $16, $26),
color set 2 as greens ($0A, $1A, $2A), and color set 3 as blues
($02, $12, $22).  By convention, the ignored bytes match the
backdrop color.

    0F0010200F0616260F0A1A2A0F021222
    ||||||||||||||||||||||||||++++++- Color set 3
    ||||||||||||||||||||||||++------- Ignored
    ||||||||||||||||||++++++--------- Color set 2
    ||||||||||||||||++--------------- Ignored
    ||||||||||++++++----------------- Color set 1
    ||||||||++----------------------- Ignored
    ||++++++------------------------- Color set 0
    ++------------------------------- Backdrop color

An activity's screenshot is 7 rows of 8 tiles, each 8 by 8 pixels.
Each tile can use one of two color sets in the palette.  Each color
set has three colors plus four grays ($0F, $00, $10, and $20).
If a screenshot fails to meet this restriction, the builder will
fail and print a diagnostic.  If no screenshot is assigned to an
activity, the builder assigns a fallback image:
`../tilesets/screenshots/default.png`

Prerequisites
-------------
To build a collection, you'll need Python 3 and Pillow (Python
Imaging Library).  To rebuild the menu itself, you'll also need
GNU Make, GNU Coreutils, cc65, and optionally the SoX audio
converter.  Instructions to install most of this are at
https://github.com/pinobatch/nrom-template

Use
---
After you've installed Python and Pillow, make a config file
and put it in a folder with your ROMs.  Then run this command,
specifying the filename of the config file and ROM file:

    tools/a53build.py example.cfg example.nes

If no fatal errors occurred, a53games.nes should appear in the
top level folder.

**Warning: The following mechanism is broken.**  
The package also includes a tool to extract ROMs from the collection.
If you have an a53games.nes file, you can extract ROMs, screenshots,
and the skeleton of a.cfg file by running a53extract.py.  Thus, a
collection is an "aggregate" under the GNU General Public License
in the same way that a bootable disc image is.  One caveat is that
the only prgunused entries in the resulting configuration file
correspond to the reset patches.

The configuration file
----------------------
The configuration file uses a name-value pair syntax similar to that
of .ini files but allowing multiple-line values.  See the docstring
at the top of tools/innie.py for detailed syntax information.

A sample config file with one activity follows:

    # BEGIN example.cfg
    
    [title]
    titlescreen=title_screen.png
    titlepalette=0f0010200f1616160f1616160f161616
    
    text=Hi Mom!
    at=64,192
    color=2,0
    
    [games]
    page=Platform
    title=Wrecking Ball Boy
    author=Justin Bailey
    year=2011
    description:
    Use your grappling hook
    to swing over gaps and
    through obstacles.

    + Move
    A: Fire or retract
      grappling hook
    .
    players=1-2 alt
    screenshot=screenshots/Wrecking Ball Boy.png
    rom=roms/Wrecking Ball Boy.nes
    prgunused3=f700-fff9

    # END example.cfg

Let's take that apart, with comments this time:

    [title]

Commands related to the title screen must appear in this section.

    titlescreen=../tilesets/title_screen.png

The title screen is a 256x240 pixel image as described above.
This and other paths are relative to the directory containing
the config file.

    titlepalette=0f0010200f1616160f1616160f161616

Sets the palette for converting the title screen to an NES
background image.
    
    text=Hi Mom!

Adds a line of text.  Multiple-line values are allowed, but
behavior is undefined if any line is longer than 128 pixels,
which is about 28 characters.

    at=64,192

Sets the top left X,Y coordinate for the latest line of text.
X is pixel precise, but Y is rounded down to a multiple of 8.

    color=2,0

Sets the foreground color for the latest line of text to color 2
and the background color to 0.  One of the colors must be 0 or 3,
and the other must be 1 or 2.  They appear with whatever attribute
was already at that position in the PNG.

    [games]

Commands related to activities must appear in this section.

    page=Platform

Begins a new page called "Platform".  The page need only
be specified once, before the first activity on the page.
If you try to put more than 20 activities on a page, the
builder will print a diagnostic.

    title=Wrecking Ball Boy

Begins a new activity with this name.

    author=Justin Bailey

The author or copyright owner of the activity.

    year=2011

The year in which the activity was first published.

    description:
    Use your grappling hook
    to swing over gaps and
    through obstacles.

    + Move
    A: Fire or retract
      grappling hook
    .

Up to 16 lines, each about 25 to 28 characters, describing
the basics of how to play a game.
A colon instead of an equal sign starts a multiple-line value,
and a dot on its own line ends it.  The first dot on any line is
stripped off.  (This escaping method is the same used in SMTP.)

    players=1-2 alt

How many players can play at once.  Acceptable values are
`1`, `2`, `1-2`, `1-2 alt`, `1-3`, `1-4`, `2-4 alt`, `2-6 alt`, and
`2-4`.  `2` means the game is for two players only, like
"Fire Breathers" from _Action 52_.  If omitted, uses `1`.

    screenshot=screenshots/Wrecking Ball Boy.png

A 64x56 pixel screenshot of the activity as described above.

    rom=roms/Wrecking Ball Boy.nes

The path to the ROM containing the activity.

    prgunused3=f700-fff9

Ranges of bytes in the PRG bank that are unused.  The builder will
usually place an exit patch here and may fill these with CHR data,
screenshots, etc.  This MUST be specified for any 32 KiB or bigger
PRG ROM, but it need only be specified in one activity within each
bank of the PRG ROM.

The following are useful primarily for submultis, or ROMs containing
more than one activity:

    mapmode=AE

Advanced: Overrides the detected mapper with a mapper configuration
code. Bit 7 is 0 for CNROM or 1 for other, bits 5-4 control the
game's PRG ROM size (32, 64, 128, or 256 KiB), and bits 3-0 have
nearly the same meaning as on MMC1's control register.
Useful primarily for submultis.

    prgbank=0

Which 32 KiB PRG bank within the PRG ROM is associated with an
activity.  If omitted, uses the first bank for fixed-$8000 (such as
UNROM configured as mapper 180) or the last bank for other mappers.

    chrbank=0

Which 8 KiB bank is associated with an activity.
If omitted, uses the first bank (0).

    entrypoint=C000

Where the activity's code starts if the PRG bank's reset vector
doesn't point there.

    patch3=C03C:F0

One-line fixes to make an activity boot correctly.  These
apply to all activities in one PRG ROM.

    exitmethod=unrom
    exitmethod=none

These control how the game is patched to jump to the menu when
Reset is pressed.  If your game has a menu option to exit, use
`none`; otherwise, the valid options are `nrom`, `unrom`, `cnrom`,
`aorom`, `unrom`, and `unrom180`.

    exitpoint=FF00

Sets the reset vector to $FF00 instead of automatically patching the
ROM to exit to the menu.  Use this if an activity has an `entrypoint`
and is already patched to exit to the main menu on Reset.

Questions
---------
Q. I've added a header to a53menu.prg.  Why does it just hang?

This file contains the code for the menu.  It contains no activities
and thus will not work unless activities are added using a53build.py.

Q. How do I use a collection with more than a dozen or so ROMs
on a PowerPak?

The PowerPak has only enough RAM to simulate 512 KiB of PRG ROM.
You'll need to rebuild some of your activities into submultis and
split your collection into smaller collections.  Action 53 volume 3
(2017), for example, had to be played with two.

Q. How do I play games that use two Zappers?

The menu supports a Zapper in port 2 or a standard controller or
Super NES Mouse in port 1.  PowerPak users can use controller 1 to
navigate to Action 53, swap to the desired controller at the title
screen, press Reset to get the NES to recognize the newly connected
controller, and then use the menu.

Q. Why can't I use [x] controller for the menu?

The vast majority of the menu program is the work of one person,
and he has an NES, not a Famicom.  To get a new NES controller
supported in the menu, make a homebrew game supporting both that
controller and the standard controller, and notify forums.nesdev.com
that you want that game included.

Legal
-----
This manual, as well as the menu and builder, are distributed under
the following terms:

Copyright 2012-2018 Damian Yerrick

Copying and distribution of this file, with or without
modification, are permitted in any medium without royalty provided
the copyright notice and this notice are preserved in all source
code copies.  This file is offered as-is, without any warranty.

(End of terms)

This product is not endorsed by Nintendo or Active Enterprises.


CHANGES.txt
0.06wip3 (2018-09-17)
* Menu: Update Pently audio driver to 2018-08
* Menu: Change title theme song to "Attack injection" from Pently
  demo pending composition of a new piece
* Menu: Use sound effects from Pino's previous Action 53 entries
* Menu: Overlap left and right sides of tabs to allow longer titles
* Menu: Select acts as Back, like B (requested by rainwarrior)
* Menu: Page up and down tries to stay on the same line number
  instead of going to the start or end of a page
* Game loader: Load actual number of CHR banks for CNROM
* Game loader: Clear nametables and almost all RAM
* a53build: Correctly handle NROM-128 when reset and IRQ vectors are
  in separate 16K halves
* Compress descriptions with digram tree encoding (DTE)
* Move some non-ASCII characters below code unit $20 to free up code
  units for DTE
* Add Control Pad and bird emoji (for contributors' Twitter
  usernames) (requested by M_Tee)
* Compress screenshots and activity CHR data with JRoatch's Donut
  codec for better compression than PB53
* Temporarily move some data in last bank below $C000 to make up
  for bigger Pently footprint

0.06wip2 (2018-03-10)
* Menu: Compress border graphics
* Menu: Read notice strings on title screen from a list pointed to in
  the keyblock, instead of hardcoded strings with hardcoded addresses
* Menu: Rearrange modules for more efficient use of page-aligned ROM
* a53build: Title screen commands go in [title] section of cfg file;
  most other commands must be in [games]
* a53build: Add list of title screen notice strings to the last bank
* autosubmulti: Disable NMI in bottom half exit patch
* Convert README to Markdown

0.06wip1 (2018-02-03)
* quadanalyze: Fall back to wave module if SoX not installed
* quadanalyze: Don't hardcode input and output filenames
* Menu: Move controller detection before title screen, relying on
  Reset press rather than hot plugging
* Menu: Shooting Zapper offscreen changes pages
* Menu: When Zapper present, read it in 25% of frames
* Menu: Don't poll Zapper in tab bar; move sprite 0 below tab bar
* Menu: Make screenshot solid while it is changing
* Menu: Screenshots can have up to 10 colors
* Menu: New TV border by tokumaru, mostly based on Sony Trinitron
  KV-20TR23
* innie: Config file character encoding is UTF-8
* a53build: User can choose characters for code units above $80
  by editing a53charset.py

0.05wip4 (2017-04-21)
* Menu: Don't include iNES header in menu-only binary to discourage
  accidental execution (requested by rainwarrior and dustmop)
* Menu: Display error if key block is missing a screenshot
  (requested by dustmop)
* autosubmulti: Map from mirroring to map mode bit 0 was backward
  (reported by RogerBidon)
* a53build: Print exception message for each ROM that failed to load
  (requested by JRoatch)
* a53build: Make paths relative to config file's directory
  (requested by JRoatch)
* a53build: Config file can change path of menu binary
* a53build: Screenshot converter shares NES palette with savtool
* a53build: Receive name of config file on command line
* makefile: `make dist` builds and packages both the menu source and
  one or more collections
* Workaround for Info-ZIP Zip's nonzero exit status when no changes
  are needed <https://stackoverflow.com/a/19258421/2738262>
* Docs: Mention new features related to Action 53 mapper

0.05wip3 (2017-04-08)
* Menu: Integrate JRoatch's coredump tool
* Menu: Remove debug code that corrupted the background when Up is
  held (reported by Greg Caldwell)
* exitpatch: Exit patches do not set stack pointer, so that coredump
  (A+B+Reset) can see the stack pointer
* exitpatch: Fixed marking exitmethod=none banks as already patched
* a53build: 32K mappers (BNROM, AOROM) default to last PRG bank
  instead of first, working around an exit patch problem
* a53build: Load ROMs on a blank-named page but don't list them
  as titles, to allow including unreachable TCRF-bait content
* Add tools/prgunused.py to find runs of 32 or more bytes of
  value $00 or $FF in the PRG ROM of an iNES image

0.05wip2 (2017-03-09)
* Menu: Include title screen in data part of last bank (near ROM
  directory) instead of menu executable
* a53build: Set filename and palette of title screen with config
  file keywords titlescreen= and titlepalette=
* a53build: Handle grayscale mode screenshots and indexed
  screenshots with black not first
* a53build: Correctly exit-patch larger ROMs (breaks a53extract)

0.05wip1 (2017-02-25)
* Menu: Update music engine from Pently 3 to Pently current
* Menu: Update VWF engine to unrolled shift version
* Menu: Update source code file naming conventions
* Menu: Remove antialiasing from start button notice
* tools/*: Convert to Python 3
* makefile: Work around quirks in Python for Windows
* a53build: Specify unused ranges for multiple PRG banks in one
  activity
* a53build: ROMs using Action 53 mapper (28) boot as if they're
  UxROM (2)
* a53build: Refactor title validation and ROM patch parsing into
  separate subroutine
* a53build: If prgbank not given, guess last ROM bank for
  UxROM (2) and Action 53 or first for others
* a53build: Fixed several title parsing error messages
* a53build: Sort ROMs by descending PRG size before inserting them,
  fixing arbitrary mixes of C?NROM and [ABU]NROM
* autosubmulti: Don't limit input or output to one directory

0.04 (2014-11-06)
* Menu: Fixed certain combinations of tab lengths causing blank
  block in 11th line of tab display
* Menu: Bank arrow sprite positioned correctly even if tab width
  differs from first tab's width
* Menu: Fixed uncleared line caused by rapidly switching from a
  long to medium to short page
* Menu: Performs a write sequence to fully initialize mapper 28
  without losing support for mapper 34
* Menu: Can customize the title screen using a .png file and a
  palette specified in the makefile
* a53build: startbank= option to leave several 32K banks blank
  at the start of the ROM, for adding mapper 28 UNROM games

src/main.s
;
; Action 53 title screen and game loader
; Copyright 2011-2014 Damian Yerrick
;
; Copying and distribution of this file, with or without
; modification, are permitted in any medium without royalty provided
; the copyright notice and this notice are preserved in all source
; code copies.  This file is offered as-is, without any warranty.
;

.include "nes.inc"
.include "global.inc"
.include "pently.inc"

.import interbank_fetch, interbank_fetch_buf
.import coredump
.importzp lineImgBufLen
.export pently_zptemp
;.export PB53_outbuf
;.exportzp ciSrc, ciBufStart, ciBufEnd

; ld65 manual says LOWCODE is for stuff guaranteed not to be
; banked out, good for irq handlers, but it's also good for
; anything that must access data from multiple banks
.import __LOWCODE_RUN__, __LOWCODE_LOAD__
.import __LOWCODE_SIZE__

OAM = $0200

.segment "ZEROPAGE"
nmis:          .res 1
oam_used:      .res 1

; Used by pads.s
cur_keys:  .res 2
new_keys:  .res 2
cur_trigger: .res 1

; Used by music engine
tvSystem:   .res 1
music_nmis: .res 1
pently_zptemp: .res 5

; Used by serial byte streams
ciSrc: .res 2
ciBufStart: .res 1
ciBufEnd: .res 1
PB53_outbuf = $0100

.segment "BSS"
linebuf: .res 40
ciBlocksLeft = linebuf

.if 0
.segment "INESHDR"
  .byt "NES",$1A  ; magic signature
  .byt 2          ; PRG ROM size in 16384 byte units
  .byt 0          ; CHR ROM size in 8192 byte units
  .byt $00        ; mirroring type and mapper number lower nibble
  .byt $00        ; mapper number upper nibble
.endif

.segment "BFF0"
.proc patch_bff0
  sei
ldxinstr:
  ldx #$FF
  nop
  stx ldxinstr+1
  jmp ($FFFC)
.endproc

.segment "FFF0"
.proc patch_fff0
  sei
ldxinstr:
  ldx #$FF
  nop
  stx ldxinstr+1
  jmp ($FFFC)
  .addr nmi, reset, irq
.endproc

; Add this only when copying hex into the patcher
.if 0
.segment "CODE"
.proc patch21
  sei
  ldx #copied_end - copied - 1
:
  lda copied,x
  sta $00,x
  dex
  bpl :-
  jmp $0000
copied:
  stx copied+5
  jmp ($FFFC)
copied_end:
.endproc
.endif

.segment "CODE"
;;
; This NMI handler is good enough for a simple "has NMI occurred?"
; vblank-detect loop, like in Lawn Mower or Thwaite.
.proc nmi
  inc nmis
  rti
.endproc

; Action 53 dosn't use IRQ. Use this to catch runaway code that
; hits a BRK opcode. This is also currently used to test the
; full build by having the dummy ROM image simply execute
; BRK while mapped to the menu.
.proc irq
  pha
  txa
  pha
  tya
  pha
jmp coredump
.endproc

; 
.proc reset
  ; The very first thing to do when powering on is to put all sources
  ; of interrupts into a known state.  But the coredump feature
  ; requires all this to be done without writing to memory or the
  ; stack pointer.
  sei             ; Disable interrupts
  ldx #$00
  stx PPUCTRL     ; Disable NMI and set VRAM increment to 32
  stx PPUMASK     ; Disable rendering
  stx $4010       ; Disable DMC IRQ
  dex             ; Leave for later TXS
  bit PPUSTATUS   ; Acknowledge stray vblank NMI across reset
  bit SNDCHN      ; Acknowledge DMC IRQ
  lda #$40
  sta P2          ; Disable APU Frame IRQ
  lda #$0F
  sta SNDCHN      ; Disable DMC playback, initialize other channels

vwait1:
  bit PPUSTATUS   ; It takes one full frame for the PPU to become
  bpl vwait1      ; stable.  Wait for the first frame's vblank.

coredump_at_boot_readpad:
  ldy #$01
  lda #$00
  sty $4016
  sta $4016
  @readpad_loop:
    lda $4017
    ora $4016
    and #%00000011  ; ignore D2-D7
    cmp #1          ; CLC if A=0, SEC if A>=1
    tya
    rol
    tay
  bcc @readpad_loop
  ; A = Y = pad2
  cpy #%11000000
  bne @not_coredump
    jmp coredump
  @not_coredump:

  ; Now that we know we're not running coredump:
  txs  ; Set stack pointer
  cld  ; Turn off decimal mode for post-patent famiclones

  ; Copy the CHR decompression code to RAM.
  .assert __LOWCODE_SIZE__ < 256, error, "LOWCODE too big"
  ldx #0
copy_LOWCODE:
  lda __LOWCODE_LOAD__,x
  sta __LOWCODE_RUN__,x
  inx
  cpx #<__LOWCODE_SIZE__
  bne copy_LOWCODE

  jsr init_mapper

  ; Wait for the second vblank and figure out what TV system we're on.
  ; After this, use only NMI to wait for vblank.
  jsr getTVSystem
  sta tvSystem
  lda #VBLANK_NMI
  sta PPUCTRL

  ; Blank the screen, wait for vblank, and wait a bit more,
  ; so that the Zapper's photodiode is detected as dark during
  ; the entire detection routine.
  lda #$3F
  ldy #$00
  sta PPUADDR
  sty PPUADDR
  sta PPUDATA
  sty PPUADDR
  sty PPUADDR
  jsr ppu_wait_vblank
  jsr pently_init
  jsr identify_controllers

  ; Prime the controller reading, so that holding a button during
  ; power-on doesn't close the title screen.
  jsr read_zapper_trigger
  jsr read_pads

  ; If the title screen pointer is in $FF00-$FFFF, no games are
  ; loaded.  This could happen if the user burns the menu bank
  ; by itself without building a ROM.
  ldy TITLESCREEN+1
  iny
  bne has_games
    jmp no_games_error
  has_games:

  jsr title_screen
  lda nmis
  sta music_nmis
  jsr cart_menu

  ; game id is in A
  pha
  jsr get_titledir_a
  jsr load_titledir_chr_rom
  ldx #0
  jsr ppu_clear_oam  ; Clear OAM because Snail Maze Game doesn't
  pla

  ; CHR data is loaded.
titleptr = $00
  jsr get_titledir_a  ; refetch because CHR ROM loading clobbered it
  ldy #TITLE_ENTRY_POINT
  lda (titleptr),y
  sta start_entrypoint+0
  iny
  lda (titleptr),y
  sta start_entrypoint+1
  iny
  .assert TITLE_ENTRY_POINT + 2 = TITLE_MAPPER_CFG, error, "title dir field order changed"
  lda (titleptr),y
  sta start_mappercfg

  ldx #0
  stx SNDCHN
  jmp start_game
.endproc

.importzp ciSrc0, ciSrc1
.import interbank_fetch, interbank_fetch_buf

; Temporary measure during debugging to keep Sinking Feeling-only ROM
; from reading off the end.  Once this is working, I'll try adding
; the CHR ROM size to the title directory.
CNROM_MAX_SIZE = 4

;;
; Loads the CHR bank ID associated with this title.
; @param $0000 pointer to entry in title directory
; @param $FF02 pointer to start of CHR directory, where each entry is
; 5 bytes: PRG bank, address low, high, midpoint offset low, high
.proc load_titledir_chr_rom
titleptr = $00
chrdir_entry_zp = $02  ; matches interbank_fetch bank ptr
chrdir_entry = interbank_fetch_buf+72
cur_chr_bank = interbank_fetch_buf+74
num_chr_banks = interbank_fetch_buf+75

  lda #1
  sta num_chr_banks
  lsr a
  sta cur_chr_bank
  sta PPUMASK
  sta PPUCTRL
  ldy #TITLE_CHR_BANK
  lda (titleptr),y
  bpl is_chr_rom  ; CHR bank >= $80 means absent
    rts
  is_chr_rom:

  ; A = starting CHR bank number, 0 to 127.
  ; Calculate offset into chrdir as 5 * A + CHRDIR_START
  .if ::TITLE_CHR_BANK <> 1
    ldy #1
  .endif
  dey
  sty chrdir_entry_zp+1
  sta chrdir_entry_zp
  asl a
  asl a
  rol chrdir_entry_zp+1
  adc chrdir_entry_zp
  bcc :+
    inc chrdir_entry_zp+1
    clc
  :
  adc CHRDIR_START
  sta chrdir_entry
  lda CHRDIR_START+1
  adc chrdir_entry_zp+1
  sta chrdir_entry+1

  ldy #TITLE_NUMBER_OF_CHR
  lda (titleptr),y
  sta num_chr_banks
nextbank:
  lda #$00
  sta $5000
  lda cur_chr_bank
  sta $8000
  inc cur_chr_bank
  lda #$81
  sta $5000
  lda chrdir_entry
  sta chrdir_entry_zp
  lda chrdir_entry+1
  sta chrdir_entry_zp+1
  ldy #0
  sty draw_progress
  iny
  lda (chrdir_entry_zp),y
  iny
  sta ciSrc0
  lda (chrdir_entry_zp),y
  iny
  sta ciSrc0+1
  clc
  lda (chrdir_entry_zp),y
  iny
  adc ciSrc0
  sta ciSrc1
  lda (chrdir_entry_zp),y
  adc ciSrc0+1
  sta ciSrc1+1

  ; At this point:
  ; 2: pointer to bank number
  ; ciSrc0: Address of first 4 KiB
  ; ciSrc1: Address of second 4 KiB
  ; drawProgress: number of blocks already drawn
loop:
  lda draw_progress
  lsr a
  ror 0
  lsr a
  ror 0
  sta PPUADDR
  lda 0
  and #$c0
  sta PPUADDR

  lda ciSrc0
  sta 0
  lda ciSrc0+1
  sta 1
  jsr do4
  adc ciSrc0
  sta ciSrc0
  bcc :+
  inc ciSrc0+1
:

  lda draw_progress
  lsr a
  ror 0
  lsr a
  ror 0
  eor #$10
  sta PPUADDR
  lda 0
  and #$c0
  sta PPUADDR

  lda ciSrc1
  sta 0
  lda ciSrc1+1
  sta 1
  jsr do4
  adc ciSrc1
  sta ciSrc1
  bcc :+
  inc ciSrc1+1
:

  inc draw_progress
  lda draw_progress
  cmp #64
  bcc loop

  dec num_chr_banks
  beq no_more_chr_banks
    clc
    lda #5
    adc chrdir_entry
    sta chrdir_entry
    bcc :+
      inc chrdir_entry+1
    :
    jmp nextbank
  no_more_chr_banks:
  rts

;;
; @param A start of buffer (0 or 64)
; @return A number of compressed bytes read
do4:
  lda #65  ; block of 4 tiles, max 64 + 1 bytes
  sta 4
  lda chrdir_entry
  sta chrdir_entry_zp
  lda chrdir_entry+1
  sta chrdir_entry_zp+1
  lda #0
  sta PPUCTRL
  bit PPUSTATUS
  jsr interbank_fetch
  lda #VBLANK_NMI
  sta PPUCTRL
  ldy #<interbank_fetch_buf
  lda #>interbank_fetch_buf
  ldx #1
  jsr donut_block_ayx
  jsr pently_update_lag
  lda donut_stream_ptr+0
  sec
  sbc #<interbank_fetch_buf
  clc
  rts
.endproc

;;
; Reads a mouse and maps presses of L to A and R to B
; in new_keys.
; @param X port number of mouse to read (0 or 1)
.proc read_mouse_with_backward_buttons
  jsr read_mouse
  lda new_mbuttons,x
  and #KEY_B|KEY_A
  asl a  ; bit 7: LMB; carry: RMB
  bcc :+
    ora #KEY_B
  :
  ora new_keys,x
  sta new_keys,x

  lda #0
no_mouse:
  rts
.endproc

;;
; @return nonzero if there was a 1-to-0 transition on the Zapper trigger
.proc read_zapper_trigger
  lda detected_pads
  and #DETECT_2P_ZAPPER
  beq no_zapper
    lda $4017
    and #$10
    tay
    eor #$10
    and cur_trigger
    sty cur_trigger
  no_zapper:
  rts
.endproc

.proc pently_update_lag
  lda nmis
  cmp music_nmis
  beq caught_up
    jsr pently_update
    inc music_nmis
    jmp pently_update_lag
  caught_up:
  rts
.endproc

; historic a53 volumes
; Emulator: http://www.fceux.com/web/download.html
; Volume 1: https://forums.nesdev.com/viewtopic.php?p=109907#p109907
; Volume 2: https://forums.nesdev.com/viewtopic.php?p=148343#p148343
; Volume 3 (work in progress): https://forums.nesdev.com/download/file.php?id=8115
